File: cpp/Letter-Combinations-Of-A-Phone-Number/Letter-Combinations-Of-A-Phone-Number.cpp
/// Source : https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/
/// Author : bryce
/// Time   : 2019-11-08

/*
用迭代来求解，遍历digits中所有的数字，建立一个临时的字符串数组temp,然后通过数字到dict中取出字符串str，然后遍历取出字符串所有字符，
再遍历当前结果res中的每一个字符串，将字符加到后面，并加入到临时字符串数组temp中。取出的字符串str遍历完成后，将临时字符串数组赋值给结果res
*/


// C++ Solution 1:
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.empty()) return res;
        string dict[] = {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        res.push_back("");
        for (int i = 0; i < digits.size(); i++)
        {
            int n = res.size();
            string str = dict[digits[i] - '2'];
            for (int j = 0; j < n;j++)
            {
                string temp = res.front();    //vec.front() 返回第一个元素的值，与begin()函数有区别，begin()函数返回的是第一个元素的迭代器
                res.erase(res.begin());      //删除第一个元素
                for (int k = 0; k < str.size();k++)
                {
                    string a = temp + str[k];
                    res.push_back(a);
                }
            }
        }
        return res;      
    }
};

// C++ Solution 2:
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char,string> mp{{'2',"abc"},{'3',"def"},{'4',"ghi"},{'5',"jkl"},
                                      {'6',"mno"},{'7',"pqrs"},{'8',"tuv"},{'9',"wxyz"}};
        vector<string> result{""};
        if(digits.size()==0) return {};
        for(int i=0;i<digits.size();i++){
            vector<string> temp;
            string str = mp[digits[i]];
            for(int j=0;j<str.size();j++){
                for(string s:result)          // C++11 里面的for的新写法，表示依次取出字符串数组result中的元素
                    temp.push_back(s+str[j]);
            }
            result = temp;
        }
        return result;
    }
};

/*
其实可以将此题看成求解一棵树的所有root(root可以看做是空)到叶子的路径。  
例如当`digits = "23"`时，树应该是这个样子:
```
          root
        /   |   \
  2:   a    b    c
      /|\  /|\  /|\
  3:  def  def  def
```
所以就可以用DFS求解这题了。
*/

File: python/Letter-Combinations-Of-A-Phone-Number/Letter-Combinations-Of-A-Phone-Number.py
#! /usr/bin/env python
#! -*- coding=utf-8 -*-
# Date: 2019-11-08
# Author: Bryce


### Python Solution1:
class Solution(object):
    def letterCombinations(self, digits):
        """
        动态规划
        dp[i]: 前i个字母的所有组合
        由于dp[i]只与dp[i-1]有关,可以使用变量代替列表存储降低空间复杂度
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        d = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
             '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
        n = len(digits)
        dp = [[] for _ in range(n)]
        dp[0] = [x for x in d[digits[0]]]
        for i in range(1, n):
            dp[i] = [x + y for x in dp[i - 1] for y in d[digits[i]]]
        return dp[-1]

      
### Python Solution2:
class Solution(object):    
    def letterCombinations(self, digits):
        """
        使用变量代替上面的列表
        降低空间复杂度
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        dic = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
             '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
        n = len(digits)
        res = ['']
        for i in range(n):
            res = [x + y for x in res for y in dic[digits[i]]]
        return res

      
###  Python Solution3:
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        dic = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
             '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
        if not digits:
            return []
        if len(digits) == 1:
            return [x for x in dic[digits[0]]]
        return [x + y for x in dic[digits[0]] for y in self.letterCombinations(digits[1:])]
